lib/auth/init.go

```diff
@@ -484,6 +483,49 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
 	return asrv, nil
 }
 
+func initSetAuthPreference(asrv *Server, newAuthPref services.AuthPreference) error {
+	storedAuthPref, err := asrv.GetAuthPreference()
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		storedAuthPref = nil
+	}
+	shouldReplace, err := shouldInitReplaceResource(storedAuthPref, newAuthPref)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if shouldReplace {
+		err = asrv.SetAuthPreference(newAuthPref)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		log.Infof("Updating auth preference: %v.", newAuthPref)
+	}
+	return nil
+}
+
+// shouldInitReplaceResource determines whether the candidate resource should
+// be used to replace the stored resource during auth server initialization.
+//
+// For more details, see Section "Auth server initialization"
+// in rfd/0016-dynamic-configuration.md
+func shouldInitReplaceResource(stored, candidate types.ResourceWithOrigin) (bool, error) {
+	if candidate.Origin() != types.OriginConfigFile && candidate.Origin() != types.OriginDefaults {
```

Comment:
Nit: `Check()` method on the Origin type may be helpful.

Reply:
Also, can `candidate` be nil here?

Reply:
The `nil` case should be excluded here although it could occur by a programmer mistake. `candidate` is what comes from the default configuration which is updated with what's specified in the config file. I'll add a `nil` check just to be sure.